src/gz/bufread.rs

```diff
@@ -280,108 +298,213 @@ impl<R: BufRead + Write> Write for GzEncoder<R> {
-            | ((buf[1] as u32) << 8)
-            | ((buf[2] as u32) << 16)
-            | ((buf[3] as u32) << 24);
-        let amt = ((buf[4] as u32) << 0)
-            | ((buf[5] as u32) << 8)
-            | ((buf[6] as u32) << 16)
-            | ((buf[7] as u32) << 24);
-        if crc != self.inner.crc().sum() {
-            return Err(corrupt());
-        }
-        if amt != self.inner.crc().amount() {
-            return Err(corrupt());
-        }
-        self.finished = true;
-        Ok(())
+    fn multi(mut self, flag: bool) -> GzDecoder<R> {
+        self.multi = flag;
+        self
     }
 }
 
 impl<R> GzDecoder<R> {
     /// Returns the header associated with this stream, if it was valid
     pub fn header(&self) -> Option<&GzHeader> {
-        self.header.as_ref().and_then(|h| h.as_ref().ok())
+        self.header.as_ref()
     }
 
     /// Acquires a reference to the underlying reader.
     pub fn get_ref(&self) -> &R {
-        self.inner.get_ref().get_ref()
+        self.reader.get_ref().get_ref()
     }
 
     /// Acquires a mutable reference to the underlying stream.
     ///
     /// Note that mutation of the stream may result in surprising results if
     /// this encoder is continued to be used.
     pub fn get_mut(&mut self) -> &mut R {
-        self.inner.get_mut().get_mut()
+        self.reader.get_mut().get_mut()
     }
 
     /// Consumes this decoder, returning the underlying reader.
     pub fn into_inner(self) -> R {
-        self.inner.into_inner().into_inner()
+        self.reader.into_inner().into_inner()
     }
 }
 
 impl<R: BufRead> Read for GzDecoder<R> {
     fn read(&mut self, into: &mut [u8]) -> io::Result<usize> {
-        match self.header {
-            None => return Ok(0), // error already returned,
-            Some(Ok(_)) => {}
-            Some(Err(_)) => match self.header.take().unwrap() {
-                Ok(_) => panic!(),
-                Err(e) => return Err(e),
-            },
+        let GzDecoder { inner, header, reader, multi } = self;
+
+        enum Next {
+            None,
+            Header,
+            Body,
+            Finished,
+            Err(io::Error),
+            End
         }
-        if into.is_empty() {
-            return Ok(0);
-        }
-        match self.inner.read(into)? {
-            0 => {
-                self.finish()?;
-                Ok(0)
+
+        let mut next = Next::None;
+
+        loop {
+            match inner {
+                GzState::Header(buf) => {
+                    let mut reader = Buffer::new(buf, reader.get_mut().get_mut());
+                    match read_gz_header(&mut reader) {
+                        Ok(hdr) => {
+                            *header = Some(hdr);
+                            next = Next::Body;
+                        },
+                        Err(ref err) if io::ErrorKind::WouldBlock == err.kind()
+                            || io::ErrorKind::UnexpectedEof == err.kind()
```

Comment:
Out of curiosity, how come `UnexpectedEof` is here in addition to `WouldBlock`? I'd expect that `UnexpectedEof` would be a "completely fatal" error we'd want to return rather than handle

Reply:
You are right... `UnexpectedEof` should not appear.